pkg/scheduler/framework/plugins/nodeunschedulable/node_unschedulable.go

```diff
@@ -48,10 +50,45 @@ const (
 // failed by this plugin schedulable.
 func (pl *NodeUnschedulable) EventsToRegister() []framework.ClusterEventWithHint {
 	return []framework.ClusterEventWithHint{
-		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.UpdateNodeTaint}},
+		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add | framework.UpdateNodeTaint}, QueueingHintFn: pl.isSchedulableAfterNodeChange},
 	}
 }
 
+// isSchedulableAfterNodeChange is invoked for all node events reported by
+// an informer. It checks whether that change made a previously unschedulable
+// pod schedulable.
+func (pl *NodeUnschedulable) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	originalNode, modifiedNode, err := util.As[*v1.Node](oldObj, newObj)
+	if err != nil {
+		logger.Error(err, "unexpected objects in isSchedulableAfterNodeChange", "oldObj", oldObj, "newObj", newObj)
+		return framework.QueueAfterBackoff
+	}
+
+	podToleratesUnschedulable := v1helper.TolerationsTolerateTaint(pod.Spec.Tolerations, &v1.Taint{
+		Key:    v1.TaintNodeUnschedulable,
+		Effect: v1.TaintEffectNoSchedule,
+	})
+
+	if podToleratesUnschedulable {
+		// Basically, shouldn't reach here
+		// because Pod.Spec.Tolerations is an immutable field and this plugin never rejects Pod with Toleration for Unschedulable taint.
+		return framework.QueueAfterBackoff
+	}
+
+	originalNodeSchedulable, modifiedNodeSchedulable := false, !modifiedNode.Spec.Unschedulable
```

Comment:
I think we don't need to care about the `originalNodeSchedulable` for it's always false, if it's true, we'll pass the nodeUnschedulable plugin, which makes the logic simpler.

Reply:
We need it.
Let's say NodeA.Spec.Unschedulable=false and NodeB.Spec.Unschedulable=true, and Pod is failed by this plugin.
In this case, in this`isSchedulableAfterNodeChange`, we should ignore all changes to NodeA because NodeA is not related to this plugin. What we need to care about is the change to NodeB only.

So, in order to filter out such non-related events for NodeA, we need to have `originalNodeSchedulable` to return QueueAfterBackoff only when NodeB.Spec.Unschedulable=true → NodeB.Spec.Unschedulable=false.

Reply:
why would be return QueueAfterBackoff if nodeB.Spec.Unchedulable=true?

Reply:
I mean QueueAfterBackoff should be returned only when NodeB gets changed from Unschedulable=true to Unschedulable=false.

Reply:
Got it, thanks for the explanation.

Reply:
I still don't understand why we need to restrict returning QueueAfterBackoff to the transition.

I get that, in general, it shouldn't matter: once there is a transition to Unschedulable=false, the plugin would only return Success, so this function shouldn't be called.

But before #120334, we would call this function again. I think it's actually safer not to restrict passing this check to transitions only.

Reply:
Let's say there is only one Node with `Unschedulable=true` in the cluster but almost all other Nodes are `Unschedulable=false`. In that case, all Pods would get NodeUnschedulable in the unschedulable plugins.
And, if we didn't restrict returning QueueAfterBackoff to the transition and returned `QueueAfterBackoff` to all `Unschedulable=false` Nodes, we would keep requeueing rejected Pods to activeQ/backoffQ with tons of unrelated events to unrelated Nodes (`Unschedulable=false` from the first). What we should observe is the event to the Node which is/was `Unschedulable=true` only.

> But before https://github.com/kubernetes/kubernetes/pull/120334, we would call this function again. I think it's actually safer not to restrict passing this check to transitions only.

So... yes. If we want to consider unknown bugs like https://github.com/kubernetes/kubernetes/pull/120334 which aren't found yet, we should make all plugins to be more conservative, returning QueueAfterBackoff not only to the transition. But, I'm not sure if it's worth doing for unknown bugs. 
Like NodeUnschedulable, many Node level scheduling constraint (like NodeTaint, NodeAffinity...etc), the same thing would happen -- like we need to return QueueAfterBackoff to changes to all untainted Nodes, we need to return QueueAfterBackoff to changes to all match Nodes, etc.